ouroboros-consensus/src/Ouroboros/Storage/ImmutableDB/Impl/Validation.hs

```diff
@@ -373,6 +358,38 @@ validateEpoch ValidateEnv{..} shouldBeFinalised epoch mbPrevHash = do
         UnexpectedError (e@InvalidFileError {}) -> Just e
         _ -> Nothing) err
 
+    -- | When reading the entries from the secondary index file, we need to
+    -- pass in a value of type 'IsEBB' so we know whether the first entry
+    -- corresponds to an EBB or a regular block. We need this information to
+    -- correctly interpret the deserialised 'Word64' as a 'BlockOrEBB': if
+    -- it's an EBB, it is the 'EpochNo' ('Word64'), it's a regular block, is a
+    -- 'SlotNo' ('Word64').
+    --
+    -- However, at the point we are reading the secondary index file, we don't
+    -- yet know whether the first block will be an EBB or a regular block. We
+    -- will find that out when we read the actual block from the epoch file.
+    --
+    -- Fortunately, we can make a /very/ good guess: if the 'Word64' of the
+    -- 'BlockOrEBB' matches the epoch number, it is almost certainly an EBB,
+    -- as the slot numbers increase @10k@ times faster than epoch numbers.
+    -- E.g., the first slot number in epoch 1 will be 21600 if @k = 2160@. We
+    -- could only make the wrong guess in the first very first epoch, i.e.,
+    -- epoch 0, as the first slot number is also 0. However, we know that the
+    -- real blockchain starts with an EBB, so even in that case we're fine.
+    --
+    -- If the epoch size were 1, then we would make the wrong guess for each
+    -- epoch that contains an EBB, which is a rather unrealistic scenario.
+    --
+    -- Note that even making the wrong guess is not a problem. The only thing
+    -- that doesn't go according to plan is that we will needlessly overwrite
+    -- the secondary index file.
```

Comment:
This is a bit subtle, and could do with some additional documentation: we extract _only_ the CRC from the index file, pass that to the parser, and then _reconstruct_ the entire index file and rewrite if there is a mismatch. This isn't totally obvious; for example, it wouldn't be crazy to also pass to the parser whether or not we expect the block to be an EBB (after all, this information is present in the index file we just read). That would break the logic here.